File: src/core/Options/Handlers/OptionTransactionHandler.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using MediatR;

namespace core.Options
{
    public class OptionTransactionHandler : 
        MediatR.INotificationHandler<OptionSold>,
        MediatR.INotificationHandler<OptionPurchased>
    {
        private IPortfolioStorage _storage;
        private IMediator _mediator;

        public OptionTransactionHandler(IPortfolioStorage storage, IMediator mediator)
        {
            _storage = storage;
            _mediator = mediator;
        }

        public async Task Handle(OptionSold e, CancellationToken cancellationToken)
        {
            var o = await _storage.GetOwnedOption(e.AggregateId, e.UserId);
            if (o == null)
            {
                return;
            }

            var when = e.When;
            var notes = e.Notes;
            var ticker = o.State.Ticker;

            await CreateNote(e.UserId, when, notes, ticker);
        }

        public async Task Handle(OptionPurchased e, CancellationToken cancellationToken)
        {
            var o = await _storage.GetOwnedOption(e.AggregateId, e.UserId);
            if (o == null)
            {
                return;
            }

            var when = e.When;
            var notes = e.Notes;
            var ticker = o.State.Ticker;

            await CreateNote(e.UserId, when, notes, ticker);
        }

        private async Task CreateNote(Guid userId, DateTimeOffset when, string notes, string ticker)
        {
            if (string.IsNullOrEmpty(notes))
            {
                return;
            }
            
            var cmd = new Notes.Add.Command
            {
                Created = when,
                Note = notes,
                Ticker = ticker
            };

            cmd.WithUserId(userId);

            await _mediator.Send(cmd);
        }
    }
}

File: src/core.fs/Options/OptionTransactionHandler.fs
namespace core.fs.Options

    open core
    open core.Options
    open core.fs

    type OptionTransactionHandler(storage:IPortfolioStorage) =
        
        let createNote userId ``when`` notes (ticker:string) = task {
            
            match notes with
            | null -> return ()
            | "" -> return ()
            | _ ->
                let note = core.Notes.Note(userId=userId, ``created``=``when``, note=notes, ticker=ticker)
                let! _ = storage.Save(note, userId)
                return ()
        }
        
        interface IApplicationService
            
        member _.HandleSell (sell:OptionSold) = task {
            
            let! o = storage.GetOwnedOption(sell.AggregateId, sell.UserId)
            
            match o with
            | null ->
                return ()
            | _ ->
                let ``when`` = sell.When
                let notes = sell.Notes
                let! _ = createNote sell.UserId ``when`` notes (o.State.Ticker)
                return ()
        }
        
        member _.HandleBuy (buy:OptionPurchased) = task {
            
            let! o = storage.GetOwnedOption(buy.AggregateId, buy.UserId)
            
            match o with
            | null ->
                return ()
            | _ ->
                let ``when`` = buy.When
                let notes = buy.Notes
                let! _ = createNote buy.UserId ``when`` notes (o.State.Ticker)
                return ()
        }
            